File: mdix-java/src/main/test/java/com/midmanstudio/dixscript/SchemaTest.java
package com.midmanstudio.dixscript;

import org.junit.jupiter.api.*;
import static org.assertj.core.api.Assertions.*;

/**
 * Integration tests for {@link SchemaBuilder}.
 * Requires the native lib to be on java.library.path (set by build.gradle.kts).
 */
class SchemaTest {

    private static final String VALID_SRC =
        "@DATA( app_name = \"MyApp\" port = 8080 debug = true )";

    private static final String MISSING_FIELD_SRC =
        "@DATA( port = 8080 )";

    private static final String WRONG_TYPE_SRC =
        "@DATA( app_name = \"MyApp\" port = \"not-a-number\" )";

    // ── passing validation ────────────────────────────────────────────────────

    @Test void validate_allFieldsPresentAndTyped_passes() {
        try (Database db = DixScript.loadStr(VALID_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireString("app_name")
                .requireInt("port")
                .optionalBool("debug")
                .validate(db);
            assertThat(report.isValid()).isTrue();
            assertThat(report.errorCount()).isZero();
        }
    }

    @Test void validate_missingOptionalField_stillPasses() {
        try (Database db = DixScript.loadStr(VALID_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireString("app_name")
                .optionalString("not_present")
                .validate(db);
            assertThat(report.isValid()).isTrue();
        }
    }

    // ── failing validation ────────────────────────────────────────────────────

    @Test void validate_missingRequiredField_reportsMissing() {
        try (Database db = DixScript.loadStr(MISSING_FIELD_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireString("app_name")
                .requireInt("port")
                .validate(db);
            assertThat(report.isValid()).isFalse();
            assertThat(report.errorsOfKind(SchemaBuilder.ErrorKind.MISSING)).hasSize(1);
            assertThat(report.failedPaths()).containsExactly("app_name");
        }
    }

    @Test void validate_wrongType_reportsWrongType() {
        try (Database db = DixScript.loadStr(WRONG_TYPE_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireString("app_name")
                .requireInt("port")
                .validate(db);
            assertThat(report.isValid()).isFalse();
            assertThat(report.errorsOfKind(SchemaBuilder.ErrorKind.WRONG_TYPE)).hasSize(1);
        }
    }

    @Test void validate_multipleErrors_reportsAll() {
        try (Database db = DixScript.loadStr(MISSING_FIELD_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireString("app_name")
                .requireString("author")
                .validate(db);
            assertThat(report.errorCount()).isEqualTo(2);
        }
    }

    // ── custom validators (requireWith / optionalWith) ───────────────────────

    @Test void requireWith_customValidatorPasses() {
        try (Database db = DixScript.loadStr(VALID_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireWith("port", SchemaBuilder.ExpectedType.INT, data -> {
                    int port = data.getInt("port");
                    return (port >= 1025 && port <= 65535) ? null : "port out of range";
                })
                .validate(db);
            assertThat(report.isValid()).isTrue();
        }
    }

    @Test void requireWith_customValidatorFails_reportsInvalidValue() {
        try (Database db = DixScript.loadStr(VALID_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireWith("port", SchemaBuilder.ExpectedType.INT, data -> "always fails")
                .validate(db);
            assertThat(report.isValid()).isFalse();
            assertThat(report.errorsOfKind(SchemaBuilder.ErrorKind.INVALID_VALUE)).hasSize(1);
        }
    }

    @Test void requireWith_skipsCustomValidator_whenTypeCheckAlreadyFailed() {
        try (Database db = DixScript.loadStr(MISSING_FIELD_SRC)) {
            boolean[] validatorRan = { false };
            SchemaBuilder.Report report = new SchemaBuilder()
                .requireWith("app_name", SchemaBuilder.ExpectedType.STRING, data -> {
                    validatorRan[0] = true;
                    return null;
                })
                .validate(db);
            assertThat(report.isValid()).isFalse();
            assertThat(validatorRan[0]).isFalse(); // never ran — the field was missing (type check failed first)
        }
    }

    // ── metadata ──────────────────────────────────────────────────────────────

    @Test void fieldCount_and_paths_reflectAddedFields() {
        SchemaBuilder schema = new SchemaBuilder().requireString("a").requireInt("b").optionalBool("c");
        assertThat(schema.fieldCount()).isEqualTo(3);
        assertThat(schema.paths()).containsExactly("a", "b", "c");
    }

    @Test void withDescription_beforeAnyField_throws() {
        assertThatThrownBy(() -> new SchemaBuilder().withDescription("oops")).isInstanceOf(MdixException.class);
    }

    @Test void report_toString_isHumanReadable() {
        try (Database db = DixScript.loadStr(MISSING_FIELD_SRC)) {
            SchemaBuilder.Report report = new SchemaBuilder().requireString("app_name").validate(db);
            assertThat(report.toString()).contains("Validation failed").contains("app_name");
        }
    }
}

File: mdix-php/tests/SchemaTest.php
<?php
declare(strict_types=1);

namespace MidManStudio\Mdix\Tests;

use MidManStudio\Mdix\ExpectedType;
use MidManStudio\Mdix\MdixDatabase;
use MidManStudio\Mdix\MdixError;
use MidManStudio\Mdix\MdixSchemaBuilder;
use MidManStudio\Mdix\ValidationErrorKind;
use PHPUnit\Framework\TestCase;

/**
 * Integration tests for MdixSchemaBuilder.
 * Requires the native lib to be on MDIX_LIB_PATH or in mdix-php/lib/.
 */
final class SchemaTest extends TestCase
{
    private const VALID_SRC = '@DATA( app_name = "MyApp" port = 8080 debug = true )';
    private const MISSING_FIELD_SRC = '@DATA( port = 8080 )';
    private const WRONG_TYPE_SRC = '@DATA( app_name = "MyApp" port = "not-a-number" )';

    public function testValidateAllFieldsPresentAndTypedPasses(): void
    {
        $db = MdixDatabase::loadStr(self::VALID_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireString('app_name')
                ->requireInt('port')
                ->optionalBool('debug')
                ->validate($db);

            self::assertTrue($report->isValid());
            self::assertSame(0, $report->errorCount());
        } finally {
            $db->close();
        }
    }

    public function testValidateMissingOptionalFieldStillPasses(): void
    {
        $db = MdixDatabase::loadStr(self::VALID_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireString('app_name')
                ->optionalString('not_present')
                ->validate($db);

            self::assertTrue($report->isValid());
        } finally {
            $db->close();
        }
    }

    public function testValidateMissingRequiredFieldReportsMissing(): void
    {
        $db = MdixDatabase::loadStr(self::MISSING_FIELD_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireString('app_name')
                ->requireInt('port')
                ->validate($db);

            self::assertFalse($report->isValid());
            self::assertCount(1, $report->errorsOfKind(ValidationErrorKind::Missing));
            self::assertSame(['app_name'], $report->failedPaths());
        } finally {
            $db->close();
        }
    }

    public function testValidateWrongTypeReportsWrongType(): void
    {
        $db = MdixDatabase::loadStr(self::WRONG_TYPE_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireString('app_name')
                ->requireInt('port')
                ->validate($db);

            self::assertFalse($report->isValid());
            self::assertCount(1, $report->errorsOfKind(ValidationErrorKind::WrongType));
        } finally {
            $db->close();
        }
    }

    public function testValidateMultipleErrorsReportsAll(): void
    {
        $db = MdixDatabase::loadStr(self::MISSING_FIELD_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireString('app_name')
                ->requireString('author')
                ->validate($db);

            self::assertSame(2, $report->errorCount());
        } finally {
            $db->close();
        }
    }

    public function testRequireWithCustomValidatorPasses(): void
    {
        $db = MdixDatabase::loadStr(self::VALID_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireWith('port', ExpectedType::Int, function (MdixDatabase $data): ?string {
                    $port = $data->getInt('port');
                    return ($port >= 1025 && $port <= 65535) ? null : 'port out of range';
                })
                ->validate($db);

            self::assertTrue($report->isValid());
        } finally {
            $db->close();
        }
    }

    public function testRequireWithCustomValidatorFailsReportsInvalidValue(): void
    {
        $db = MdixDatabase::loadStr(self::VALID_SRC);
        try {
            $report = (new MdixSchemaBuilder())
                ->requireWith('port', ExpectedType::Int, fn (MdixDatabase $data): string => 'always fails')
                ->validate($db);

            self::assertFalse($report->isValid());
            self::assertCount(1, $report->errorsOfKind(ValidationErrorKind::InvalidValue));
        } finally {
            $db->close();
        }
    }

    public function testRequireWithSkipsCustomValidatorWhenTypeCheckAlreadyFailed(): void
    {
        $db = MdixDatabase::loadStr(self::MISSING_FIELD_SRC);
        try {
            $validatorRan = false;
            $report = (new MdixSchemaBuilder())
                ->requireWith('app_name', ExpectedType::String, function (MdixDatabase $data) use (&$validatorRan): ?string {
                    $validatorRan = true;
                    return null;
                })
                ->validate($db);

            self::assertFalse($report->isValid());
            self::assertFalse($validatorRan); // never ran -- the field was missing (type check failed first)
        } finally {
            $db->close();
        }
    }

    public function testFieldCountAndPathsReflectAddedFields(): void
    {
        $schema = (new MdixSchemaBuilder())->requireString('a')->requireInt('b')->optionalBool('c');
        self::assertSame(3, $schema->fieldCount());
        self::assertSame(['a', 'b', 'c'], $schema->paths());
    }

    public function testWithDescriptionBeforeAnyFieldThrows(): void
    {
        $this->expectException(MdixError::class);
        (new MdixSchemaBuilder())->withDescription('oops');
    }

    public function testReportToStringIsHumanReadable(): void
    {
        $db = MdixDatabase::loadStr(self::MISSING_FIELD_SRC);
        try {
            $report = (new MdixSchemaBuilder())->requireString('app_name')->validate($db);
            $text = (string) $report;
            self::assertStringContainsString('Validation failed', $text);
            self::assertStringContainsString('app_name', $text);
        } finally {
            $db->close();
        }
    }
}
